pipit/util/test_generator.py

```diff
@@ -0,0 +1,366 @@
+import random
+import textwrap
+import pandas as pd
+
+
+class FakeNode:
+    """
+    A single node of the calling tree used to produce fake traces.
+    Represents a single function call.
+    """
+
+    def __init__(
+        self, name, exc_time, mpi_type="", mpi_tgt=0, mpi_volume=0, mpi_time=0
+    ):
+        self.name = name
+        self.exc_time = exc_time
+        self.inc_time = exc_time
+        self.children = {}  # run_time -> child node
+        self.is_mpi = mpi_type != ""
+        self.mpi_type = mpi_type
+        self.mpi_tgt = mpi_tgt
+        self.mpi_volume = mpi_volume
+        self.mpi_time = mpi_time
+        self.total_nodes = 1
+        self.parent = None
+
+    def grow_inc_time(self, time):
+        """
+        This function adjusts inclusive time metric when a new child is added,
+        adding the time to each parent's inc_time, up to the root.
+        """
+        self.inc_time += time
+        if self.parent is not None:
+            self.parent.grow_inc_time(time)
+
+    def grow_total_nodes(self, amt):
+        """
+        This function adjusts the total count of nodes when a new node is added,
+        adding the new count to each parent's total_nodes, up to the root.
+        """
+        self.total_nodes += amt
+        if self.parent is not None:
+            self.parent.grow_total_nodes(amt)
+
+    def add_child(self, child, run_time):
+        """
+        This function adds a child that executes after run_time
+        exclusive time within the function represented by the current node.
+        """
+        assert run_time not in self.children
```

Comment:
do we also want to assert that two children don't overlap or is this not possible currently?

Reply:
The way it is structured, children can't overlap each other, since `run_time` is _exclusive_ time in that current node. Even if a child has a very long inclusive time and the `run_time` difference between it and the next one is small, the next one will still be ordered after it.

Reply:
Thanks, this makes sense. I was confused about what exactly `run_time` is, but that clears it up now.